pymtl3/datatypes/bit_structs.py

```diff
@@ -0,0 +1,442 @@
+  r : Bits4
+  g : Bits4
+  b : Bits4 = b4(5)
+
+  def __str__( self ):
+    return f'({self.r},{self.g},{self.b})'
+
+Author : Yanghui Ou
+  Date : July 25, 2019
+"""
+import keyword
+import types
+import warnings
+
+#-------------------------------------------------------------------------
+# Errors
+#-------------------------------------------------------------------------
+
+class NoFieldDeclaredError( AttributeError ):
+  ...
+
+#-------------------------------------------------------------------------
+# Constants
+#-------------------------------------------------------------------------
+
+# Object with this attribute is considered as bit struct, as we assume
+# only the bit_struct decorator will stamp this attribute to a class. This
+# attribute also stores the field information and can be used for
+# translation.
+_IS_BIT_STRUCT = '__bit_struct_fields__'
+
+_DEFAULT_SELF_NAME = 's'
+_ANTI_CONFLICT_SELF_NAME = '__bit_struct_self__'
+
+
+# A sentinel object to detect whether a parameter is present or not. Here
+# we use a class to give a better repr.
+class _MISSING_TYPE:
+  ...
+MISSING = _MISSING_TYPE()
+
+#-------------------------------------------------------------------------
+# Field
+#-------------------------------------------------------------------------
+# Field instances are used for holding the name, type, and default value
+# of each field of the bit struct.
+
+class Field:
+  # Here we use __slots__ to declare date members to potentially save
+  # space and improve look up time.
+  __slots__ = (
+    'name',
+    'type_',
+    'default',
+  )
+
+  def __init__( self, name=None, type_=None, default=MISSING ):
+    self.name    = name
+    self.type_   = type_
+    self.default = default
+
+  def __repr__( self ):
+    return (
+      'Field('
+      f'name={self.name!r},'
+      f'type_={self.type_!r},'
+      f'default={self.default!r}'
+      ')'
+    )
+
+#-------------------------------------------------------------------------
+# field
+#-------------------------------------------------------------------------
+# Public APIs for creating a Field object. This is used for supporting
+# dictionary-like syntax for mk_bit_struct like
+#
+# mk_bit_struct( 'Point',{
+#   'x' : field( Bits4 )
+#   'y' : field( Bits4, default=b4(f) )
+# }
+
+def field( type_, default=MISSING ):
+  return Field( None, type_, default )
+
+#-------------------------------------------------------------------------
+# _create_fn
+#-------------------------------------------------------------------------
+# A helper function that creates a function based on
+# - fn_name : name of the function
+# - args_lst : a list of arguments in string
+# - body_lst : a list of statement of the function body in string
+#
+# Note: this function mutates _locals and should only be called internally
+# to this module.
+
+def _create_fn( fn_name, args_lst, body_lst, _globals=None, _locals=None ):
+
+  # Lazily construct empty dictionary
+  if _locals is None:
```

Comment:
what about _globals?

Reply:
Oh I think I did this because as I tried `exec` can take globals as none but not locals.

Reply:
if locals is None then exec will mutate the current locals()

Reply:
ok!

Reply:
maybe add some comments on this exec thing here